File: 200. Number of Islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:

        ans = 0
        visited = set()

        def bfs(i, j):

            q = collections.deque()
            q.append((i, j))

            # top, bottom, left, right
            moveDir = [(-1, 0), (1, 0), (0, -1), (0, 1)]

            while q:
                current = q.popleft()
                visited.add(current)
                for move in moveDir:
                    moveI, moveJ = current[0] + move[0], current[1] + move[1]
                    if ((moveI, moveJ) not in visited and
                            moveI in range(len(grid)) and
                            moveJ in range(len(grid[moveI])) and
                            grid[moveI][moveJ] == "1"):
                        q.append((moveI, moveJ))
                        visited.add((moveI, moveJ))


        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if grid[i][j] == "1" and (i, j) not in visited:
                    bfs(i, j)
                    ans += 1
                visited.add((i, j))

        return ans

File: 200. Number of Islands.java
class Solution {

    private void bfs(char[][] grid, int i, int j, boolean[][] visited) {
        // top, bottom, left, right
        int[][] directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

        int[] startPoint = {i, j};
        Queue<int[]> q = new ArrayDeque<>();
        q.add(startPoint);

        while (!q.isEmpty()) {
            //System.out.println("----");
            int[] currentPoint = q.remove();
            i = currentPoint[0];
            j = currentPoint[1];
            //System.out.println("running " + i + " " + j);
            for (int[] d: directions) {
                int movedI = i+d[0];
                int movedJ = j+d[1];
                if (movedI >= 0 && movedI < grid.length && movedJ >= 0 &&
                    movedJ < grid[movedI].length &&
                    !visited[movedI][movedJ] &&
                    grid[movedI][movedJ] == '1') {
                    int[] newPoint = {movedI, movedJ};
                    q.add(newPoint);
                    visited[movedI][movedJ] = true;
                }
            }
        }
    }

    public int numIslands(char[][] grid) {
        boolean[][] visited = new boolean[grid.length][grid[0].length];
        int ans = 0;

        for (int i=0; i<grid.length; i++) {
            for (int j=0; j<grid[i].length; j++) {
                if (grid[i][j] == '1' && !visited[i][j]) {
                    bfs(grid, i, j, visited);
                    ans += 1;
                }
            }
        }
        return ans;
    }
}